### PR TITLE
gh-121999: Change default tarfile filter to 'data'

### PR DESCRIPTION
**Pull request by WilliamRoyNelson**
* Issue: gh-121999

----
📚 Documentation preview 📚: https://cpython-previews--122002.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/tarfile.rst [hunk 1]
```diff
@@ -40,9 +40,11 @@ Some facts and figures:
    Archives are extracted using a :ref:`filter <tarfile-extraction-filter>`,
    which makes it possible to either limit surprising/dangerous features,
    or to acknowledge that they are expected and the archive is fully trusted.
-   By default, archives are fully trusted, but this default is deprecated
-   and slated to change in Python 3.14.
 
+.. versionchanged:: 3.14
+   The default extraction filter was 'fully trusted' but is now 'data' which
```
Comment (tomasr8) [inline Doc/library/tarfile.rst]: Could be useful to add a reference to the filters here:
```suggestion
   The default extraction filter was :func:`fully_trusted <fully_trusted_filter>` but is now :func:`data <data_filter>` which
```

File: Lib/test/test_tarfile.py [hunk 2]
```diff
@@ -738,31 +738,6 @@ def test_extract_directory(self):
         finally:
             os_helper.rmtree(DIR)
 
-    def test_deprecation_if_no_filter_passed_to_extractall(self):
```
Comment (tomasr8) [inline Lib/test/test_tarfile.py:741]: A new test case to check that the default is `data_filter` would also be nice :)

**Comment by ZeroIntensity** [comment]
CC @encukou, as this is PEP 706.

File: Lib/test/test_tarfile.py (hunk 2 above)
Comment (WilliamRoyNelson) [reply Lib/test/test_tarfile.py:741]: My buddy @sodle just added one:
https://github.com/python/cpython/pull/122002/commits/15216b9aa19aeb3d316c416b64139637bb54dd75

File: Doc/library/shutil.rst [hunk 3]
```diff
@@ -706,17 +706,18 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
...
-   in Python 3.14.
 
    .. audit-event:: shutil.unpack_archive filename,extract_dir,format shutil.unpack_archive
 
    .. warning::
 
-      Never extract archives from untrusted sources without prior inspection.
+      Never extract archives from untrusted sources without prior inspection,
+      even when using the ``'data'`` filter, but especially if using the
+      ``'tar'`` or ``'fully_trusted'`` filters.
```
Comment (picnixz) [inline Doc/library/shutil.rst]: There are only three filters for now and I think that an unstrusted source englobes the entire source, whether some parts are trusted or not. 

I mean, if I use 'fully_trusted', it contradicts the fact that it's assumed to be untrusted =/ (In summary, the current wording is fine to me).

File: Doc/library/tarfile.rst [hunk 4]
```diff
@@ -40,9 +40,11 @@ Some facts and figures:
    Archives are extracted using a :ref:`filter <tarfile-extraction-filter>`,
    which makes it possible to either limit surprising/dangerous features,
    or to acknowledge that they are expected and the archive is fully trusted.
-   By default, archives are fully trusted, but this default is deprecated
-   and slated to change in Python 3.14.
 
+.. versionchanged:: 3.14
+   The default extraction filter was :func:`fully_trusted <fully_trusted_filter>` but is now :func:`data <data_filter>` which
+   which disallows dangerous features like links to absolute paths or paths
+   outside the destination.
```
Comment (picnixz) [inline Doc/library/tarfile.rst]:
```suggestion
   Set the default extraction filter to :func:`data <data_filter>`,
   which disallows dangerous features such as links to absolute paths
   or paths outside of the destination. Previously, the filter strategy
   was :func:`fully_trusted <fully_trusted_filter>`.
```

File: Doc/library/tarfile.rst [hunk 5]
```diff
@@ -495,19 +497,23 @@ be finalized; only the internally used file object will be closed. See the
...
-   you need to support.
+   It is recommended to set this explicitly only if unusual *tar* features
+   are required.
 
    .. warning::
 
-      Never extract archives from untrusted sources without prior inspection.
+      The default filter is set to ``filter='data'`` to prevent the most
+      dangerous security issues, read the :ref:`tarfile-extraction-filter`
+      section for details.
```
Comment (picnixz) [inline Doc/library/tarfile.rst]:
```suggestion
      The default filter is set to ``filter='data'`` to prevent the most
      dangerous security issues. Read the :ref:`tarfile-extraction-filter`
      section for details.
```

File: Doc/library/tarfile.rst [hunk 6]
```diff
@@ -538,8 +544,9 @@ be finalized; only the internally used file object will be closed. See the
 
       See the warning for :meth:`extractall`.
 
-      Set ``filter='data'`` to prevent the most dangerous security issues,
-      and read the :ref:`tarfile-extraction-filter` section for details.
+      The default filter is set to ``filter='data'`` to prevent the most
+      dangerous security issues, read the :ref:`tarfile-extraction-filter`
```
Comment (picnixz) [inline Doc/library/tarfile.rst]:
```suggestion
      dangerous security issues. Read the :ref:`tarfile-extraction-filter`
```

File: Lib/test/test_tarfile.py [hunk 7]
```diff
@@ -722,6 +722,28 @@ def format_mtime(mtime):
             tar.close()
             os_helper.rmtree(DIR)
 
+    @staticmethod
+    @unittest.mock.patch("tarfile.data_filter", wraps=tarfile.data_filter)
+    @unittest.mock.patch("tarfile.tar_filter", wraps=tarfile.tar_filter)
+    @unittest.mock.patch("tarfile.fully_trusted_filter", wraps=tarfile.fully_trusted_filter)
+    def test_extractall_default_filter(mock_ft_filter: unittest.mock.Mock,
+                                       mock_tar_filter: unittest.mock.Mock,
+                                       mock_data_filter: unittest.mock.Mock):
```
Comment (picnixz) [inline Lib/test/test_tarfile.py]: Just use a regular method + chained with statements. Easier to read and to understand.

File: Lib/test/test_tarfile.py [hunk 8]
```diff
@@ -722,6 +722,28 @@ def format_mtime(mtime):
...
+        tar = tarfile.open(tarname, encoding="iso8859-1")
+        DIR = os.path.join(TEMPDIR, "extractall_default_filter")
+        os.mkdir(DIR)
+        try:
```
Comment (picnixz) [inline Lib/test/test_tarfile.py]: Why not use a context manager as in the previous tests for deprecation?

File: Misc/NEWS.d/next/Core_and_Builtins/2024-07-18-21-19-04.gh-issue-121999.8IBbTK.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Update tarfile library to use 'data' filter by default when extracting
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-07-18-21-19-04.gh-issue-121999.8IBbTK.rst]:
```suggestion
The default extraction filter for the :mod:`tarfile` module is now
set to :func:`'data' <tarfile.data_filter>`.
```

File: Lib/test/test_tarfile.py (hunk 7 above)
Comment (sodle) [reply Lib/test/test_tarfile.py]: Roger, will fix.

File: Lib/test/test_tarfile.py (hunk 8 above)
Comment (sodle) [reply Lib/test/test_tarfile.py]: Ah, didn't notice that. I copied the test above this one (`test_extractall`) which uses the try/finally pattern. I'll update both tests to use the context manager.

File: Doc/library/shutil.rst (hunk 3 above)
Comment (WilliamRoyNelson) [reply Doc/library/shutil.rst]: Good point, I made an update here:
57c60e7287823fd34d8e1365e40c63d34ca2fb40

I think this gets the point across without being awkward.

File: Lib/test/test_tarfile.py (hunk 8 above)
Comment (picnixz) [reply Lib/test/test_tarfile.py]: Just modify your test, do not touch others unless it's not related. That way the diff is smaller. 

In the future, we might want to just refactor all tests but that's bordering a cosmetic PR.

File: Doc/library/shutil.rst (hunk 3 above)
Comment (picnixz) [reply Doc/library/shutil.rst]: Yes, that's better!

**Comment by picnixz** [comment]
I'm a bit confused but... arae there two people working on this PR simultaneously @WilliamRoyNelson and @sodle?

File: Lib/test/test_tarfile.py [hunk 10]
```diff
@@ -694,10 +694,11 @@ def test_extract_hardlink(self):
...
         # and times (see issue1735).
-        tar = tarfile.open(tarname, encoding="iso8859-1")
-        DIR = os.path.join(TEMPDIR, "extractall")
-        os.mkdir(DIR)
-        try:
+        DIR = pathlib.Path(TEMPDIR) / "extractall"
+        with (
+            os_helper.temp_dir(DIR),
+            tarfile.open(tarname, encoding="iso8859-1") as tar
+        ):
```
Comment (picnixz) [inline Lib/test/test_tarfile.py]: If it's not related to the PR itself, I think you can just revert this change (that way, your diff is smaller).

File: Doc/library/tarfile.rst [hunk 11]
```diff
@@ -40,9 +40,12 @@ Some facts and figures:
    Archives are extracted using a :ref:`filter <tarfile-extraction-filter>`,
    which makes it possible to either limit surprising/dangerous features,
    or to acknowledge that they are expected and the archive is fully trusted.
-   By default, archives are fully trusted, but this default is deprecated
-   and slated to change in Python 3.14.
 
+.. versionchanged:: 3.14
+   Set the default extraction filter to :func:`data <data_filter>`,
+   which disallows dangerous features such as links to absolute paths
+   or paths outside of the destination. Previously, the filter strategy
```
Comment (picnixz) [inline Doc/library/tarfile.rst:47]: Sorry about this one but the `outside of` now feels weird to me :') 

@AA-Turner As a native speaker (you're the only one I know...), should it be "outside the destination", or "outside of"? (or something else entirely?)

**Review by picnixz — COMMENTED** [review]
Some minor comments.

**Comment by encukou** [comment]
I'll review later this week.

File: Lib/test/test_tarfile.py (hunk 8 above)
Comment (sodle) [reply Lib/test/test_tarfile.py]: Makes sense. Reverted.

File: Doc/library/tarfile.rst (hunk 11 above)
Comment (WilliamRoyNelson) [reply Doc/library/tarfile.rst:47]: I think "outside of" is conventional in North America, less conventional in the UK.
https://learningenglish.voanews.com/a/should-we-think-outside-or-outside-of-the-box-/6434530.html

I think I was reading from PEP 706 when I wrote that update.
> Refuse to extract links (hard or soft) which end up linking to a path outside of the destination. (On systems that don’t support links, tarfile will, in most cases, fall back to creating regular files. This proposal doesn’t change that behaviour.)

**Comment by encukou** [comment]
For the documentation, communicating via GitHub review comments wouldn't be effective, so I took the liberty of pushing a commit to this PR directly. I hope you don't mind.

The main themes are:
- Passing `filter='data'` is still recommended, if there's any chance the code will run on Python 3.13 or below. When 3.14 is released, that's nearly all the code :)
- All the security notes should lead to `:ref:`tarfile-extraction-filter` (to explain things) and/or `:ref:`tarfile-further-verification` (to explain that `'data'` still isn't unconditionally “safe”). Hopefully we can do that without annoying people who read those already :)

For shutil: zipfile also has some safeties, though they haven't been reviewed in a while. IMO we can claim for both formats that the defaults “prevent the most dangerous of such security issues”.

**Comment by encukou** [comment]
Thank you for the update!
